File: Strings/URLParser.cpp
How it works:
Regex match splits URL into components.

parse() extracts parts from regex capture groups.

parseQueryParams() splits query string by &, then by = into a map.

Public getters give access to URL parts and query parameters.

setURL() allows replacing the URL, re-validating, and re-parsing.

Why hybrid?
Regex ensures only valid URL formats are accepted.

Manual parsing lets you work with parts easily.

Query params extracted cleanly into a map for quick lookup.


#include <iostream>
#include <string>
#include <regex>
#include <map>
#include <sstream>
#include <stdexcept>
#include <algorithm>
#include <cctype>
#include <iterator>

class URL {
private:
    std::string url;
    std::string scheme, host, path, query, fragment;
    int port = -1; // -1 means no port specified
    std::map<std::string, std::string> query_params;

    // Regex to validate URL and capture components
    static const std::regex url_regex;

    bool isValid(const std::string& url_str) const {
        return std::regex_match(url_str, url_regex);
    }

    void parse() {
        std::smatch match;
        if (!std::regex_match(url, match, url_regex)) {
            throw std::invalid_argument("Invalid URL format");
        }

        // match groups:
        // 1: scheme
        // 2: host (domain or IP)
        // 3: port (optional)
        // 4: path (optional)
        // 5: query (optional, includes '?')
        // 6: fragment (optional, includes '#')

        scheme = match[1];
        host = match[2];

        std::string port_str = match[3];
        if (!port_str.empty()) {
            port = std::stoi(port_str.substr(1)); // skip ':'
        } else {
            port = -1;
        }

        path = match[4];
        query = match[5];
        if (!query.empty() && query[0] == '?')
            query.erase(query.begin()); // remove '?'

        fragment = match[6];
        if (!fragment.empty() && fragment[0] == '#')
            fragment.erase(fragment.begin()); // remove '#'

        parseQueryParams();
    }

    void parseQueryParams() {
        query_params.clear();
        std::istringstream qstream(query);
        std::string kv;
        while (std::getline(qstream, kv, '&')) {
            if (kv.empty()) continue;
            auto pos = kv.find('=');
            if (pos == std::string::npos) {
                query_params[kv] = ""; // key with no value
            } else {
                std::string key = kv.substr(0, pos);
                std::string value = kv.substr(pos + 1);
                query_params[key] = value;
            }
        }
    }

public:
    URL(const std::string& url_str) : url(url_str) {
        if (!isValid(url)) {
            throw std::invalid_argument("Invalid URL format");
        }
        parse();
    }

    void setURL(const std::string& new_url) {
        if (!isValid(new_url)) {
            throw std::invalid_argument("Invalid URL format");
        }
        url = new_url;
        parse();
    }

    bool validate() const {
        return isValid(url);
    }

    std::string getScheme() const { return scheme; }
    std::string getHost() const { return host; }
    int getPort() const { return port; }
    std::string getPath() const { return path; }
    std::string getQuery() const { return query; }
    std::string getFragment() const { return fragment; }
    std::map<std::string, std::string> getQueryParams() const { return query_params; }

    std::string getURL() const { return url; }
};

// Initialize the regex outside the class
const std::regex URL::url_regex(
    R"(^(https?|ftp):\/\/)"                         // scheme
    R"(([a-zA-Z0-9\-\.]+|\[[0-9a-fA-F:]+\]|\d{1,3}(\.\d{1,3}){3}))" // host (domain or IPv6 or IPv4)
    R"((:\d+)?)"                                    // port (optional)
    R"((\/[^\s?#]*)?)"                              // path (optional)
    R"((\?[^\s#]*)?)"                               // query (optional)
    R"((#[^\s]*)?$)",                               // fragment (optional)
    std::regex::icase);

int main() {
    try {
        URL url("https://example.com:8080/path/to/resource?foo=bar&baz=qux#frag");
        std::cout << "Scheme: " << url.getScheme() << "\n";
        std::cout << "Host: " << url.getHost() << "\n";
        std::cout << "Port: " << url.getPort() << "\n";
        std::cout << "Path: " << url.getPath() << "\n";
        std::cout << "Query: " << url.getQuery() << "\n";
        std::cout << "Fragment: " << url.getFragment() << "\n";

        std::cout << "Query parameters:\n";
        for (const auto& [key, value] : url.getQueryParams()) {
            std::cout << "  " << key << " = " << value << "\n";
        }

        // Modify URL example
        url.setURL("ftp://192.168.0.1/resource?x=1&y=2");
        std::cout << "\nNew URL host: " << url.getHost() << "\n";
        std::cout << "New URL query params:\n";
        for (const auto& [k, v] : url.getQueryParams()) {
            std::cout << "  " << k << " = " << v << "\n";
        }

    } catch (const std::invalid_argument& e) {
        std::cerr << "Error: " << e.what() << "\n";
    }
}

File: Strings/URLParser.py
import re
from urllib.parse import unquote_plus

class URL:
    # Regex to validate URL and capture groups:
    url_regex = re.compile(
        r'^(https?|ftp)://'                          # scheme
        r'([a-zA-Z0-9\-\.]+|\[[0-9a-fA-F:]+\]|\d{1,3}(?:\.\d{1,3}){3})'  # host (domain or IPv6 or IPv4)
        r'(?::(\d+))?'                               # port (optional)
        r'(\/[^\s?#]*)?'                             # path (optional)
        r'(\?[^\s#]*)?'                              # query (optional)
        r'(#[^\s]*)?$',                              # fragment (optional)
        re.IGNORECASE
    )

    def __init__(self, url: str):
        self.url = url
        if not self.validate():
            raise ValueError("Invalid URL format")
        self._parse()

    def validate(self):
        return bool(self.url_regex.match(self.url))

    def _parse(self):
        match = self.url_regex.match(self.url)
        self.scheme = match.group(1)
        self.host = match.group(2)
        self.port = int(match.group(3)) if match.group(3) else None
        self.path = match.group(4) if match.group(4) else ''
        self.query = match.group(5)[1:] if match.group(5) else ''  # remove leading '?'
        self.fragment = match.group(6)[1:] if match.group(6) else ''  # remove leading '#'
        self.query_params = self._parse_query_params(self.query)

    def _parse_query_params(self, query):
        params = {}
        if not query:
            return params
        pairs = query.split('&')
        for pair in pairs:
            if not pair:
                continue
            if '=' in pair:
                key, value = pair.split('=', 1)
                params[unquote_plus(key)] = unquote_plus(value)
            else:
                params[unquote_plus(pair)] = ''
        return params

    def set_url(self, new_url: str):
        self.url = new_url
        if not self.validate():
            raise ValueError("Invalid URL format")
        self._parse()

    def get_scheme(self): return self.scheme
    def get_host(self): return self.host
    def get_port(self): return self.port
    def get_path(self): return self.path
    def get_query(self): return self.query
    def get_fragment(self): return self.fragment
    def get_query_params(self): return self.query_params
    def get_url(self): return self.url

# Example usage:
if __name__ == "__main__":
    try:
        url = URL("https://example.com:8080/path/to/resource?foo=bar&baz=qux#frag")
        print("Scheme:", url.get_scheme())
        print("Host:", url.get_host())
        print("Port:", url.get_port())
        print("Path:", url.get_path())
        print("Query:", url.get_query())
        print("Fragment:", url.get_fragment())
        print("Query Params:", url.get_query_params())

        url.set_url("ftp://192.168.0.1/resource?x=1&y=2")
        print("\nNew URL Host:", url.get_host())
        print("New URL Query Params:", url.get_query_params())
    except ValueError as e:
        print("Error:", e)
